File: biddo_client/lib/core/repositories/bid.dart
import 'dart:convert';
import 'package:get/get.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart' as google_map;

import '../controllers/flash.dart';
import '../models/bid.dart';
import '../services/image_compress.dart';
import 'base.dart';

class BidRepository {
  var dio = Get.find<Api>();
  var flashController = Get.find<FlashController>();
  var imageCompressService = Get.find<ImageCompressService>();

  Future<bool> delete(String bidId) async {
    try {
      await dio.api.delete('/bid/$bidId');
      return true;
    } catch (error) {
      print('Error removing bid: $error');
      return false;
    }
  }

  Future<bool> markBidsFromAuctionAsSeen(String auctionId) async {
    try {
      await dio.api.put(
        '/bid/markBidsAsSeen/$auctionId',
      );

      return true;
    } catch (error) {
      return false;
    }
  }

  Future<bool> update(
    String bidId, [
    String? rejectionReason,
    bool? isRejected,
    bool? isAccepted,
  ]) async {
    if (isRejected == null && isAccepted == null) {
      throw Exception('You must provide either isRejected or isAccepted');
    }

    try {
      await dio.api.put(
        '/bid/$bidId',
        data: jsonEncode(
          {
            "isAccepted": isAccepted,
            "isRejected": isRejected,
            "rejectionReason": rejectionReason,
          },
        ),
      );

      return true;
    } catch (error) {
      flashController.showMessageFlash(
        'Could not update bid. Please try again.',
      );
      return false;
    }
  }

  Future<Bid?> create(String auctionId, CreateBidParams params) async {
    var description = params.description!.replaceAll('\n\n\n\n\n\n', '\n');
    try {
      var result = await dio.api.post('/bid/$auctionId', data: {
        'latLng': jsonEncode(params.latLng),
        'location': params.location,
        'description': description,
        'price': params.price,
        'usedExchangeRateId': params.usedExchangeRateId,
        'initialCurrencyId': params.initialCurrencyId,
      });
      return Bid.fromJSON(result.data);
    } catch (error) {
      print('Error creating bid: $error');
      flashController
          .showMessageFlash('Could not create bid. Please try again.');
      return null;
    }
  }
}

class CreateBidParams {
  final String? description;
  final double price;
  google_map.LatLng latLng;
  final String location;

  String? usedExchangeRateId;
  String? initialCurrencyId;

  CreateBidParams({
    this.description = '',
    required this.price,
    required this.latLng,
    required this.location,
    this.usedExchangeRateId,
    this.initialCurrencyId,
  });
}

File: core/repositories/bid.ts
import { LatLng } from '../domain/auction'
import { Bid } from '../domain/bid'
import { RequestMaker, RequestType } from '../services/request-maker'

class BidRepository {
  async delete(bidId: string): Promise<boolean> {
    try {
      await RequestMaker.makeRequest({
        method: RequestType.DELETE,
        path: `/bid/${bidId}`,
      })
      return true
    } catch (error) {
      console.error('Error removing bid:', error)
      return false
    }
  }

  async markBidsFromAuctionAsSeen(auctionId: string): Promise<boolean> {
    try {
      await RequestMaker.makeRequest({
        method: RequestType.PUT,
        path: `/bid/markBidsAsSeen/${auctionId}`,
      })
      return true
    } catch (error) {
      console.error('Error marking bids as seen:', error)
      return false
    }
  }

  async update(
    bidId: string,
    rejectionReason?: string,
    isRejected?: boolean,
    isAccepted?: boolean
  ): Promise<boolean> {
    if (isRejected === undefined && isAccepted === undefined) {
      throw new Error('You must provide either isRejected or isAccepted')
    }

    try {
      await RequestMaker.makeRequest({
        method: RequestType.PUT,
        path: `/bid/${bidId}`,
        payload: JSON.stringify({
          isAccepted,
          isRejected,
          rejectionReason,
        }),
        contentType: 'application/json',
      })
      return true
    } catch (error) {
      console.error('Could not update bid:', error)
      return false
    }
  }

  async create(auctionId: string, params: CreateBidParams): Promise<Bid | null> {
    const description = params.description?.replace(/\n{3,}/g, '\n') || ''

    try {
      const result = (await RequestMaker.makeRequest({
        method: RequestType.POST,
        path: `/bid/${auctionId}`,
        payload: JSON.stringify({
          latLng: JSON.stringify(params.latLng),
          location: params.location,
          description,
          price: params.price,
          initialCurrencyId: params.initialCurrencyId,
          usedExchangeRateId: params.usedExchangeRateId,
        }),
        contentType: 'application/json',
      })) as Record<string, unknown>
      return Bid.fromJSON(result)
    } catch (error) {
      if ((error as Error).message.includes('Not enough coins')) {
        throw error
      }
      console.error('Error creating bid:', error)
      return null
    }
  }
}

export class CreateBidParams {
  description?: string
  price: number
  latLng: LatLng
  location: string
  usedExchangeRateId?: string
  initialCurrencyId?: string

  constructor(
    description: string,
    price: number,
    latLng: LatLng,
    location: string,
    usedExchangeRateId?: string,
    initialCurrencyId?: string
  ) {
    this.description = description || ''
    this.price = price
    this.latLng = latLng
    this.location = location
    this.usedExchangeRateId = usedExchangeRateId
    this.initialCurrencyId = initialCurrencyId
  }
}

const BidRepositoryInstance = new BidRepository()
export { BidRepositoryInstance as BidRepository }
